File: 5. LinkedLists/7. merge_lists/merge_lists.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

a = Node(5)
b = Node(7)
c = Node(10)
d = Node(12)
e = Node(20)
f = Node(28)
a.next = b
b.next = c
c.next = d
d.next = e
e.next = f
# 5 -> 7 -> 10 -> 12 -> 20 -> 28

q = Node(6)
r = Node(8)
s = Node(9)
t = Node(25)
q.next = r
r.next = s
s.next = t
# 6 -> 8 -> 9 -> 25

def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

# Iterative 1
def merge_lists(head1, head2):
  if(head1 is None and head2 is None):
    return None
  if(head1 is None):
    return head2
  if(head2 is None):
    return head1
  
  mergedHead = head1 if head1.val <= head2.val else head2

  prev = Node("x")
  while(head1 is not None and head2 is not None):
    if(head2.val <= head1.val):
      nexthead2 = head2.next
      prev.next = head2
      head2.next = head1
      prev = head2
      head2 = nexthead2
    else:
      prev = head1 
      head1 = head1.next
  
  if(head1 is None):
    prev.next = head2
    
  return mergedHead

# Iterative 2
def merge_lists1(head1, head2):
  dummyNode = Node(None)
  current = dummyNode

  while(head1 is not None and head2 is not None):
    if(head1.val < head2.val):
      current.next = head1
      current = current.next
      head1 = head1.next
    else:
      current.next = head2
      current = current.next
      head2 = head2.next
  
  if head1 is None:
    current.next = head2
  if head2 is None:
    current.next = head1

  return dummyNode.next


# Recursive
def merge_lists2(head1, head2):
  if(head1 is None and head2 is None):
    return None
  if(head1 is None):
    return head2
  if(head2 is None):
    return head1

  if(head1.val < head2.val):
    nextHead1 = head1.next
    head1.next = merge_lists2(nextHead1, head2)
    return head1
  else:
    nextHead2 = head2.next
    head2.next = merge_lists2(head1, nextHead2)
    return head2
  

print_list(a)
print_list(q)
print_list(merge_lists(a, q))
# print_list(merge_lists1(a, q))
# print_list(merge_lists2(a, q))
# 5 -> 6 -> 7 -> 8 -> 9 -> 10 -> 12 -> 20 -> 25 -> 28 

File: 5. LinkedLists/7. merge_lists/merge_lists.js

class Node {
    constructor(val) {
      this.val = val;
      this.next = null;
    }
}

const a = new Node(5);
const b = new Node(7);
const c = new Node(10);
const d = new Node(12);
const e = new Node(20);
const f = new Node(28);
a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;
// 5 -> 7 -> 10 -> 12 -> 20 -> 28

const q = new Node(6);
const r = new Node(8);
const s = new Node(9);
const t = new Node(25);
q.next = r;
r.next = s;
s.next = t;
// 6 -> 8 -> 9 -> 25

// Iterative
// const mergeLists = (head1, head2) => {
//     if(head1 === null && head2 === null) return null;
//     if(head1 === null) return head2;
//     if(head2 === null) return head1;
//     let dummyNode = new Node(null);
//     let current = dummyNode;
//     while(head1 !== null && head2 !== null){
//         if(head1.val < head2.val){
//             current.next = head1;
//             head1 = head1.next;
//         }else{
//             current.next = head2;
//             head2 = head2.next;
//         }
//         current = current.next;
//     }
//     if(head1 === null) current.next = head2;
//     if(head2 === null) current.next = head1;
//     return dummyNode.next;
// }


// Recursive
const mergeLists = (head1, head2) => {
    if(head1 === null && head2 === null) return null;
    if(head1 === null) return head2;
    if(head2 === null) return head1;
    if(head1.val < head2.val){
        head1.next = mergeLists(head1.next, head2);
        return head1;
    }else{
        head2.next = mergeLists(head1, head2.next);
        return head2;
    }
}    


const linkedListValues = (head, arr=[]) => {
    if(head === null) return [];
    arr = [head.val, ...linkedListValues(head.next)]
    return arr;
}

let arr1 = linkedListValues(a);
console.log(`Initial linked list 1: ${arr1}`);

let arr2 = linkedListValues(q);
console.log(`Initial linked list 2: ${arr2}`);

let newHead = mergeLists(a, q);
// 5 -> 6 -> 7 -> 8 -> 9 -> 10 -> 12 -> 20 -> 25 -> 28 

let arr3 = linkedListValues(newHead);
console.log(`Merged sorted List: ${arr3}`);


// Complexity
// Iterative
// n = length of list 1
// m = length of list 2
// Time: O(min(n, m))
// Space: O(1)

// Recursive
// n = length of list 1
// m = length of list 2
// Time: O(min(n, m))
// Space: O(min(n, m))

